src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializerHelpers.cs

```diff
@@ -0,0 +1,364 @@
+                    new List<ODataPathSegment>());
+
+                try
+                {
+                    odataPath = pathHandler.Parse(serviceRoot, id, internalRequest.RequestContainer).Path;
+                }
+                catch (ODataException)
+                {
+                    return null;
+                }
+            }
+
+            return odataPath;
+        }
+
+        internal static void ApplyODataIdContainer(object resource, ODataResourceWrapper resourceWrapper,
+            ODataDeserializerContext readContext)
+        {
+            Routing.ODataPath path = readContext.Path;
+            if (path == null)
+            {
+                return;
+            }
+
+            IEdmEntityType entityType = path.EdmType.AsElementType() as IEdmEntityType;
+
+            if (entityType != null)
+            {
+                //Setting Odataid , for POCO classes, as a property in the POCO object itself(if user has OdataIDContainer property),
+                //for Delta and EdmEntity object setting as an added property ODataIdcontianer in those classes
+
+                ODataPath ODataPath = new ODataPath(path.Segments);
+
+                // if there is no Id on the resource, try to compute one from path
+                if (resourceWrapper.ResourceBase.Id == null)
+                {
+                    ODataUri odataUri = new ODataUri { Path = ODataPath };
+                    resourceWrapper.ResourceBase.Id = odataUri.BuildUri(ODataUrlKeyDelimiter.Parentheses);
+                }
+
+                if (resourceWrapper.ResourceBase.Id != null)
+                {
+                    string odataId = resourceWrapper.ResourceBase.Id.OriginalString;
+
+                    IODataIdContainer container = new ODataIdContainer();
+                    container.ODataId = odataId;
+
+                    if (resource is EdmEntityObject edmObject)
+                    {
+                        edmObject.ODataIdContainer = container;
+                        edmObject.ODataPath = ODataPath;
+                    }
+                    else if (resource is IDeltaSetItem deltasetItem)
+                    {
+                        deltasetItem.ODataIdContainer = container;
+                        deltasetItem.ODataPath = ODataPath;
+                    }
+                    else
+                    {
+                        PropertyInfo containerPropertyInfo = EdmLibHelpers.GetClrType(entityType, readContext.Model).GetProperties().Where(x => x.PropertyType == typeof(IODataIdContainer)).FirstOrDefault();
+                        if (containerPropertyInfo != null)
+                        {
+                            IODataIdContainer odataIdContainer = containerPropertyInfo.GetValue(resource) as IODataIdContainer;
+                            if (odataIdContainer != null)
+                            {
+                                containerPropertyInfo.SetValue(resource, odataIdContainer);
```

Comment:
If you agree with my observation here, then we don't need this `if` block because we're setting back the value we just retrieved
I'd expect the `if` and `else` blocks be replaced with:
```csharp
if (odataIdContainer == null)
{
     containerPropertyInfo.SetValue(resource, container);
}
```

Reply:
The `container` and `iodatacontainer` are obtained differently.  We build the `container` and assign the `odataId` to it from the resource id.